apis/v1alpha2/httproute_types.go

```diff
@@ -0,0 +1,752 @@
+	// Hostname is the fully qualified domain name of a network host,
+	// as defined by RFC 3986. Note the following deviations from the
+	// "host" part of the URI as defined in the RFC:
+	//
+	// 1. IPs are not allowed.
+	// 2. The `:` delimiter is not respected because ports are not allowed.
+	//
+	// Incoming requests are matched against the hostnames before the
+	// HTTPRoute rules. If no hostname is specified, traffic is routed
+	// based on the HTTPRouteRules.
+	//
+	// Requests will be matched against the Host field in the following order:
+	//
+	// 1. If Hostname is precise, the request matches this rule if
+	//    the HTTP Host header is equal to the Hostname.
+	// 2. If Hostname is a wildcard, then the request matches this rule if
+	//    the HTTP Host header is to equal to the suffix
+	//    (removing the first label) of the wildcard rule.
+	// 3. If Hostname is unspecified, empty, or `*`, then any request will match
+	//    this route.
+	//
+	// If a hostname is specified by the Listener that the HTTPRoute is bound
+	// to, at least one hostname specified here must match the Listener specified
+	// hostname as per the rules above. Other hostnames will not affect processing
+	// of the route in that case.
+	//
+	// If no hostname is specified by the Listener, then that value will be treated
+	// as '*', match any hostname, and so any hostname on this Route will match.
+	//
+	// If all hostnames do not match, then the HTTPRoute is not admitted, and
+	// the implementation must raise an 'Admitted' Condition with a status of
+	// `false` for that Listener.
+	//
+	// Examples:
+	// - A Listener with unspecified, empty, or `*` values for Hostname matches
+	//   any HTTPRoute hostname.
+	// - A HTTPRoute with unspecified, empty, or `*` values for Hostname matches
+	//   any Listener hostname.
+	// - A Listener with `test.foo.com` as the hostname matches *only*
+	//   `test.foo.com` or `*.foo.com`. Any other hostnames present must be ignored.
+	// - A Listener with `*.foo.com` as hostname, all hostnames in the HTTPRoute
+	//   must have any single label where the star is, and the rest of the hostname
+	//   must match exactly. So, `test.foo.com`, `*.foo.com` or `blog.foo.com` match.
+	//   `test.blog.foo.com`, `test.bar.com`, or `bar.com` do not. Hostnames that do
+	//   not match will be ignored.
+	//
+	// Support: Core
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	Hostnames []Hostname `json:"hostnames,omitempty"`
+
+	// Rules are a list of HTTP matchers, filters and actions.
+	//
+	// +optional
+	// +kubebuilder:validation:MaxItems=16
+	// +kubebuilder:default={{matches: {{path: {type: "Prefix", value: "/"}}}}}
+	Rules []HTTPRouteRule `json:"rules,omitempty"`
+}
+
+// HTTPRouteRule defines semantics for matching an HTTP request based on
+// conditions, optionally executing additional processing steps, and forwarding
+// the request to an API object.
+type HTTPRouteRule struct {
```

Comment:
General question: url re-write is quite a common thing in L7 reverse proxies. How will users express that?

Reply:
Ended up punting on that GEP to add later, but proposal for rewrites and expanded redirects is here: https://github.com/kubernetes-sigs/gateway-api/pull/731